format/postgres/common/pg_btree/postgres/pg_btree.go

```diff
@@ -0,0 +1,284 @@
+	BTP_HAS_FULLXID      = 1 << 8 /* contains BTDeletedPageData */
+)
+
+const (
+	INDEX_SIZE_MASK       = 0x1FFF
+	INDEX_AM_RESERVED_BIT = 0x2000 /* reserved for index-AM specific usage */
+	INDEX_VAR_MASK        = 0x4000
+	INDEX_NULL_MASK       = 0x8000
+)
+
+const (
+	IndexTupleDataSize = 8
+)
+
+// struct BTMetaPageData {
+/*    0      |     4 */ // uint32 btm_magic
+/*    4      |     4 */ // uint32 btm_version
+/*    8      |     4 */ // BlockNumber btm_root
+/*   12      |     4 */ // uint32 btm_level
+/*   16      |     4 */ // BlockNumber btm_fastroot
+/*   20      |     4 */ // uint32 btm_fastlevel
+/*   24      |     4 */ // uint32 btm_last_cleanup_num_delpages
+/* XXX  4-byte hole */
+/*   32      |     8 */ // float8 btm_last_cleanup_num_heap_tuples
+/*   40      |     1 */ // _Bool btm_allequalimage
+/* XXX  7-byte padding */
+//
+/* total size (bytes):   48 */
+
+// struct BTPageOpaqueData {
+/*    0      |     4 */ // BlockNumber btpo_prev;
+/*    4      |     4 */ // BlockNumber btpo_next;
+/*    8      |     4 */ // uint32 btpo_level;
+/*   12      |     2 */ // uint16 btpo_flags;
+/*   14      |     2 */ // BTCycleId btpo_cycleid;
+//
+/* total size (bytes):   16 */
+
+// struct IndexTupleData {
+/*    0      |     6 */ // ItemPointerData t_tid;
+/*    6      |     2 */ // unsigned short t_info;
+//
+// IndexTupleData *IndexTuple;
+/* total size (bytes):    8 */
+
+func DecodePgBTree(d *decode.D, args format.PostgresBTreeIn) any {
+	btree := &BTree{
+		Args:     args,
+		PageSize: common.PageSize,
+	}
+	decodeBTreePages(btree, d)
+	return nil
+}
+
+type BTree struct {
+	Args     format.PostgresBTreeIn
+	PageSize uint64
+	page     *postgres.HeapPage
+}
+
+func decodeBTreePages(btree *BTree, d *decode.D) {
+	for i := btree.Args.Page; ; i++ {
+		page := &postgres.HeapPage{}
+		if btree.page != nil {
+			// use prev page
+			page.BytesPosBegin = btree.page.BytesPosEnd
+		}
+		page.BytesPosEnd = int64(common.TypeAlign(btree.PageSize, uint64(page.BytesPosBegin)+1))
+		btree.page = page
+
+		pos0 := page.BytesPosBegin * 8
+		d.SeekAbs(pos0)
+
+		if end, _ := d.TryEnd(); end {
+			return
+		}
+
+		if i == 0 {
+			// first page contains meta information
+			d.FieldStruct("page", func(d *decode.D) {
+				decodeBTreeMetaPage(btree, d)
+			})
+			continue
+		}
+
+		d.FieldStruct("page", func(d *decode.D) {
+			decodeBTreePage(btree, d)
+		})
+	}
+}
+
+func decodeBTreeMetaPage(btree *BTree, d *decode.D) {
+	page := btree.page
+
+	d.FieldStruct("page_header", func(d *decode.D) {
+		postgres.DecodePageHeader(page, d)
+	})
+	d.FieldStruct("meta_page_data", func(d *decode.D) {
+		decodeBTMetaPageData(d)
+	})
+
+	pos0 := d.Pos()
+	pos1 := btree.page.BytesPosSpecial * 8
+	d.FieldRawLen("unused0", pos1-pos0)
+	d.FieldStruct("page_opaque_data", func(d *decode.D) {
+		decodeBTPageOpaqueData(d)
+	})
+	pos2 := d.Pos()
+	bytesPos2 := pos2 / 8
+	if bytesPos2 != page.BytesPosEnd {
+		d.Fatalf("invalid pos after read page_opaque_data on meta page\n")
```

Comment:
You can skip `\n` in fatal end error messages

Reply:
It's done.